Support unindexed labels

avaskys opened the pull request:
I thought it would be good to keep the metricd client API as
backwards-compatible as possible. So I changed LabelSet to
be a map[string]interface{} instead of a map[string]string.
So old LabelSet literals will continue to be compatible with
the new API. The proto changes also make it so that old clients
can talk to new metricd servers although this shouldn't
matter as much.

The value of the LabelSet is either a string (old behavior:
indexed label) or an UnindexedLabel instance (new behavior:
unindexed label). All other types will throw an error.

The rest was mostly piping this through and adding tests.

Commit:
Support unindexed labels

I thought it would be good to keep the metricd client API as
backwards-compatible as possible. So I changed LabelSet to
be a map[string]interface{} instead of a map[string]string.
So old LabelSet literals will continue to be compatible with
the new API. The proto changes also make it so that old clients
can talk to new metricd servers although this shouldn't
matter as much.

The value of the LabelSet is either a string (old behavior:
indexed label) or an UnindexedLabel instance (new behavior:
unindexed label). All other types will throw an error.

The rest was mostly piping this through and adding tests.

avaskys commented on server/aggregator.go:
```diff
@@ -67,11 +66,27 @@ type aggregatorUnit interface {
+		for k := range m {
+			keys = append(keys, k)
+		}
+		// golang map iteration is in random order and we need a consistent order
```
So I wonder if the aggregation stuff actually grouped things properly before this change. I suspect that it didn't.

keunwoo replied:
nope.

keunwoo commented on pb/metrics.proto:
```diff
@@ -30,6 +37,7 @@ message Metric {
     double double_value = 11;
     string string_value = 12;
   }
+  map<string, string> unindexed_labels = 20;
```
Move this after line 31; .proto tags don't have to be in numeric order, and until protos get really giant it's not that hard to keep track of the last used tag number. It's more important to put related fields together for readability.

If you move it there, I think it makes sense to use tag number 5.

Lastly, I wonder if we should call tag 3 indexed_labels and tag 5 labels. This reads better to me than labels and unindexed_labels. The nice about binary proto's numeric tags is that the wire format is unchanged when you rename fields: you just have to change client code and recompile; you don't have to synchronously roll out schema changes to every component that touches the proto, nor do you have to execute laborious data migrations just to rename something.

avaskys replied:
I'd argue for indexed_labels and unindexed_labels. I don't like the idea of the Labels symbol in the struct suddenly meaning something else because it means I could forget to rename something and things would still compile. I'd rather force the compile error in that case.

keunwoo commented on client.go:
```diff
@@ -29,8 +29,12 @@ const normalQueueSize = 100
-type LabelSet map[string]string
+// LabelSet represents a set of metric labels.
+// Currently the supported value types are:
+//    string - the label value will be indexed in the metric database
```
This seems to encourage the default to be indexed; but if indexed labels are really a precious resource, then indexing should perhaps be opt-in rather than opt-out. Eventually I envision forcing all values in a LabelSet to be explicitly marked indexed or unindexed. But I guess we can execute that transition gradually over several CLs, and I agree that for the time being using string to mean indexed is the right choice.

The only decision we're making now that we have to live with for a long time is the naming of the UnindexedLabel type. Calling the value UnindexedLabel is a little confusing to me, because the label is the key-value pair, whereas this type just denotes the value; maybe just call it UnindexedValue, or even Unindexed, so that use sites would look like

``` go
labels := metricd.LabelSet{
    "foo": metricd.Unindexed("bar"),
}
```

avaskys replied:
Makes sense. Yeah, I like Unindexed.

keunwoo reviewed (commented):
looks fine modulo my usual quibbles about naming

Commit:
Updated naming

protobuf "labels" field --> "indexed_labels"
metricd.UnindexedLabel --> metricd.Unindexed
